dali/operators/reader/loader/webdataset_loader.cc

```diff
@@ -0,0 +1,371 @@
+  dtypes_.reserve(dtypes_ids.size());
+  std::transform(dtypes_ids.begin(), dtypes_ids.end(), std::back_inserter(dtypes_),
+                 TypeTable::GetTypeInfo);
+
+  for (auto& dtype : dtypes_) {
+    DALI_ENFORCE(detail::wds::kSupportedTypes.count(dtype.id()),
+                 make_string("Unsupported output dtype ", dtype.name(),
+                             ". Supported types include: ", SupportedTypesListGen()));
+  }
+  DALI_ENFORCE(ext_.size() == dtypes_.size(),
+               "Number of extensions does not match the number of types");
+}
+
+WebdatasetLoader::~WebdatasetLoader() {}
+
+void WebdatasetLoader::PrepareEmpty(vector<Tensor<CPUBackend>>& empty) {
+  empty = std::vector<Tensor<CPUBackend>>(ext_.size());
+  for (size_t output_index = 0; output_index < ext_.size(); output_index++) {
+    empty[output_index].reserve(tensor_init_bytes_);
+    empty[output_index].set_type(dtypes_[output_index]);
+  }
+}
+
+inline std::string GetExtension(const std::string& filepath) {
+  const size_t dot_pos = filepath.find_first_of('.', filepath.find_last_of('/') + 1);
+  return filepath.substr(dot_pos + 1);
+}
+
+void WebdatasetLoader::ReadSample(vector<Tensor<CPUBackend>>& sample) {
+  MoveToNextShard(sample_index_);
+  detail::wds::SampleDesc& current_sample = samples_[sample_index_];
+  auto& current_wds_shard = wds_shards_[current_sample.wds_shard_index];
+
+  for (auto& component : current_sample.components) {
+    current_wds_shard.SeekArchive(component.offset);
+
+    // Checking if the component data from the index file agrees with reality
+    const auto& index_path = index_paths_[current_sample.wds_shard_index];
+    DALI_ENFORCE(
+        !current_wds_shard.EndOfArchive(),
+        IndexFileErrMsg(index_path, current_sample.line_number, "archive shorter than reported"));
+    DALI_ENFORCE(
+        current_wds_shard.GetFileType() == detail::TarArchive::ENTRY_FILE,
+        IndexFileErrMsg(index_path, current_sample.line_number, "component of a non-file type"));
+    DALI_ENFORCE(GetExtension(current_wds_shard.GetFileName()) == component.ext,
+                 IndexFileErrMsg(index_path, current_sample.line_number,
+                                 "component extension does not match the archive entry extension"));
+    DALI_ENFORCE(current_wds_shard.GetFileSize() == component.size,
+                 IndexFileErrMsg(index_path, current_sample.line_number,
+                                 "component size does not match the archive entry size"));
+
+
+    // Skipping cached samples
+    const std::string source_info =
+        make_string("archive ", uris_[current_sample.wds_shard_index], "index file ",
+                    index_paths_[current_sample.wds_shard_index], "line ",
+                    current_sample.line_number, "component offset ", component.offset);
+    DALIMeta meta;
+    meta.SetSourceInfo(source_info);
+    if (ShouldSkipImage(source_info)) {
```

Comment:
I think you need to add a test case to `test_nvjpeg_cached*` inside test_pipeline.py.

Reply:
what about skip_cached_images in tes_cross_check?

Reply:
I don't think it checks it. `skip_cached_images` works only with `mixed` decoders.image, and just doesn't read any sample if it is already present in the cache of the decoder.
I think for the consistency it would be better to have this special case tested in one place.

Reply:
alright